pokemon-spring-mvc/src/main/java/cz/muni/fi/pa165/pokemon/league/participation/manager/mvc/controllers/TrainerController.java

```diff
@@ -0,0 +1,211 @@
+import cz.muni.fi.pa165.pokemon.league.participation.manager.dto.TrainerChangePasswordDTO;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.dto.TrainerRenameDTO;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.exceptions.NoSuchEntityException;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.facade.PokemonFacade;
+import cz.muni.fi.pa165.pokemon.league.participation.manager.facade.TrainerFacade;
+import javafx.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import javax.inject.Inject;
+import javax.validation.Valid;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ResourceBundle;
+
+/**
+ * Controller class for Trainer
+ *
+ * @author Michal Mokros 456442
+ */
+@Controller
+@RequestMapping("/trainer")
+public class TrainerController {
+
+    final static Logger log = LoggerFactory.getLogger(TrainerController.class);
+    ResourceBundle messages = ResourceBundle.getBundle("Texts", LocaleContextHolder.getLocale());
+
+    @Inject
+    TrainerFacade trainerFacade;
+
+    @Inject
+    PokemonFacade pokemonFacade;
+
+    /**
+     * Controller method for listing all trainers
+     * @param model data to be displayed
+     * @return JSP page
+     */
+    @RequestMapping(value = "/list", method = RequestMethod.GET)
+    public String list(Model model) {
+        log.debug("mvc list()");
+        model.addAttribute("allTrainers", trainerFacade.getAllTrainers());
+        return "trainer/list";
+    }
+
+    /**
+     * Controller method for detailed trainer
+     * @param id of the trainer
+     * @param model data to be displayed
+     * @return JSP page name
+     */
+    @RequestMapping(value = "/detail/{id}", method = RequestMethod.GET)
+    public String detail(@PathVariable long id,
+                         Model model,
+                         RedirectAttributes redirectAttributes,
+                         UriComponentsBuilder uriComponentsBuilder) {
+        log.debug("mvc detail({})", id);
+
+        if (trainerFacade.getTrainerWithId(id) == null) {
+            redirectAttributes.addFlashAttribute("alert_danger", String.format(messages.getString("trainer.does.not.exists"), id));
+            return "redirect:" + uriComponentsBuilder.path("/trainer/list").build().encode().toUriString();
+        }
+
+        model.addAttribute("trainer", trainerFacade.getTrainerWithId(id));
```

Comment:
What if trainerFacade returns null but pokemonFacade won't throw no such entity exception (timing issues: someone might have just created the trainer between the two calls)?

Reply:
So what should be the solution then? I don't really understand the scenario. If trainerFacade.getTrainerById returned null then the behavior is correct I think no matter that pokemonFacade didn't throw any exception, or are you saying I should add the exception check into if?

Reply:
Add a null check and if trainer is null then redirect to some other page (e.g. trainer/list)

Reply:
Null check around this addAttribute method that checks if the trainerFacade returns null? Just asking to be sure.

Reply:
Yes retrieve the TrainerDTO from trainerFacade, check if it's null, if so, redirect.